packages/broker/test/unit/plugins/storage/Storage.test.ts

```diff
@@ -0,0 +1,255 @@
+
+const createResultFactory = ({
+    buckets,
+    messageCounts,
+    messages
+}: {
+    buckets?: { id: string}[]|Error,
+    messageCounts?: { total: { low: number } }[]|Error,
+    messages?: { payload: string }[]|Error
+}) => {
+    return (query: string) => {
+        if (query.includes('FROM bucket')) {
+            return buckets!
+        } else if (query.includes('total FROM stream_data')) {
+            return messageCounts!
+        } else if (query.includes('payload FROM stream_data')) {
+            return messages!
+        } else {
+            throw new Error(`Assertion failed: query="${query}"`)
+        }
+    }
+}
+
+const createMockStorage = (getResults: (query: string) => any[]|Error) => {
+    const cassandraClient = {
+        eachRow: jest.fn().mockImplementation((
+            query: string,
+            _params: any[],
+            _options: any,
+            rowCallback: (n: number, row: any) => void,
+            resultCallback?: (err: Error|undefined, result: any) => void
+        ) => {
+            const rows = getResults(query)
+            if (rows instanceof Error) {
+                resultCallback!(rows, undefined)
+            } else {
+                rows.forEach((row: any, index: number) => rowCallback!(index, row))
+                resultCallback!(undefined, rows)
+            }
+        }),
+        execute: jest.fn().mockImplementation((query: string): Promise<any> => {
+            const rows = getResults(query)
+            if (rows instanceof Error) {
+                return Promise.reject(rows)
+            } else { 
+                return Promise.resolve({
+                    first: () => rows[0],
+                    rows
+                })
+            }
+        }),
+        stream: jest.fn().mockImplementation((query: string): PassThrough => {
+            const rows = getResults(query)
+            const stream = new PassThrough({
+                objectMode: true
+            })
+            if (rows instanceof Error) {
+                stream.destroy(rows)
+            } else { 
+                rows.forEach((row) => stream.write(row))
+                stream.push(null)
+            }
+            return stream
+        }),
+        shutdown: jest.fn().mockResolvedValue(undefined)
+    }
+    return new Storage(cassandraClient as any, {})
+}
+
+describe('Storage', () => {
+
+    let storage: Storage
+
+    afterEach(async () => {
+        await storage!.close()
+    })
+
+    describe('requestLast', () => {
+
+        it('happy path', (done) => {
+            storage = createMockStorage(createResultFactory({
+                buckets: [MOCK_BUCKET],
+                messageCounts: [MOCK_MESSAGE_COUNT],
+                messages: [MOCK_MESSAGE_2]
+            }))
+            const resultStream = storage.requestLast(MOCK_STREAM_ID, 0, 1)
+            const onData = jest.fn()
+            resultStream.on('data', onData)
+            resultStream.on('end', () => {
```

Comment:
IMO `done` callback should be avoided at all costs. Typically you end up missing code paths (e.g. error events) meaning if things don't occur as expected the test will just hang.

In saying that, working with async/await and streams is a bit clunky, but you can do it like this using `import { once } from 'events'`: 
```js
const results = []
resultStream.on('data', (results) => results.push(data))
await once(resultStream, 'end') // also attaches error handler
expect(results.map((r) => r.serialize())).toEqual([MOCK_MESSAGE_2.payload])
```

Reply:
Using waitForEvent and waitForStreamToEnd (see @harbu's suggestion)